services/src/main/java/org/keycloak/protocol/oidc/mappers/AbstractUserRoleMappingMapper.java

```diff
@@ -68,14 +68,21 @@ protected static void setClaim(IDToken token, ProtocolMapperModel mappingModel,
             realmRoleNames = rolesToAdd;
         }
 
-        Object claimValue = realmRoleNames;
+        // Convert the set of role names to an array
+        String[] roles = new String[realmRoleNames.size()];
+        realmRoleNames.toArray(roles);
 
         boolean multiValued = "true".equals(mappingModel.getConfig().get(ProtocolMapperUtils.MULTIVALUED));
+        Object claimValue = roles;
+
         if (!multiValued) {
-            claimValue = realmRoleNames.toString();
+            if(roles.length > 0 ) {
+                claimValue = roles[0]; // Get the first value of the array, which will be just a string value
+            } else {
+                claimValue = ""; // There were no roles available so the claim is blank
+            }
         }
```

Comment:
For the PR you wanted to update the code for the !multiValues path, yet I see lots of code changes outside of that path. Please rewrite the code so that only the multiValues path is affected.

If the there are multiple roles in the array, please suggest a way to make the result of picking one of the roles for the returned value deterministic: A set doesn't have an order, therefore the code you suggest will at worst select a random role.